h/app.py

```diff
@@ -46,6 +47,32 @@ def create_app(global_config, **settings):
     return config.make_wsgi_app()
 
 
+def publish_after_commit_events(request, response):
+    if request.exception is not None:
+        return
+
+    if len(request.after_commit_events) == 0:
+        return
+
+    with request.tm:
+        request.after_commit_events.publish_all()
+
+
+class EventQueue(collections.deque):
+    def __init__(self, request):
+        self.request = request
+
+    def publish_all(self):
+        while self.has_events():
+            event = self.popleft()
+            self.request.registry.notify(event)
+
+    def has_events(self):
+        if len(self) == 0:
+            return False
+        return True
```

Comment:
`return len(self) == 0`. But this might even be simple enough that it doesn't even need to be a separate method? Actually you don't even need `len()`, you _could_ do:

```python
while True:
    try:
        event = self.popleft()
    except IndexError:
        break
    self.request.registry.notify(event)
```

which is an "easier to ask for forgiveness than permission" approach.

(I don't think there's anything particularly wrong with the `len()` based approach though, at least in this case when th